executor/executor.go

```diff
@@ -1907,51 +1907,98 @@ func (e *TrimExec) Next() (*Row, error) {
 // UnionExec has multiple source Executors, it executes them sequentially, and do conversion to the same type
 // as source Executors may has different field type, we need to do conversion.
 type UnionExec struct {
-	schema expression.Schema
-	Srcs   []Executor
-	ctx    context.Context
-	cursor int
+	schema   expression.Schema
+	Srcs     []Executor
+	ctx      context.Context
+	inited   bool
+	finished atomic.Value
+	rowsCh   chan []*Row
+	rows     []*Row
+	cursor   int
+	wg       sync.WaitGroup
+	closedCh chan struct{}
+	errCh    chan error
 }
 
 // Schema implements the Executor Schema interface.
 func (e *UnionExec) Schema() expression.Schema {
 	return e.schema
 }
 
+func (e *UnionExec) waitAllFinished() {
+	e.wg.Wait()
+	close(e.rowsCh)
+	close(e.closedCh)
+}
+
+func (e *UnionExec) fetchData(idx int) {
+	defer e.wg.Done()
+	for {
+		rows := make([]*Row, 0, batchSize)
+		for i := 0; i < batchSize; i++ {
+			if e.finished.Load().(bool) {
+				return
+			}
+			row, err := e.Srcs[idx].Next()
+			if err != nil {
+				e.finished.Store(true)
+				e.errCh <- err
+				return
+			}
+			if row == nil {
+				if len(rows) > 0 {
+					e.rowsCh <- rows
+				}
```

Comment:
If `e.row == nil` and `len(rows) == 0` are true,  `e.rowCh` and `e.errCh` are all empty, the `select` in line 1977 is blocked.